pinc/ProofProject.inc

```diff
@@ -0,0 +1,217 @@
+<?php
+declare(strict_types=1);
+
+class ProofProjectPage extends LPage
+{
+    public function __construct(object $proof_project, object $project_page)
+    {
+        global $pguser;
+
+        $this->page_state = $project_page->page_state;
+        $this->page_name = $project_page->page_name;
+        $this->project = $proof_project->project;
+        $this->projectid = $proof_project->project->projectid;
+        $this->round = $proof_project->round;
+
+        parent::__construct($this->projectid, $this->page_name, $this->page_state, 0);
+    }
+
+    public function checkout($user)
+    {
+        // use DPage function
+        $this->page_state = Page_checkout($this->projectid, $this->imagefile, $this->round, $user);
+        return $this->render_page_data();
+    }
+
+    public function attempt_save_as_done(string $page_text): void
+    {
+        global $pguser;
+
+        $this->page_text = $this->get_valid_text($page_text);
+        if ($this->round->has_a_daily_page_limit()) {
+            $pre_save_dpl_count = get_dpl_count_for_user_in_round($pguser, $this->round);
+            if ($pre_save_dpl_count >= $this->round->daily_page_limit) {
+                // The user has already reached their limit of this kind of page.
+                $this->saveAsInProgress($page_text, $pguser);
+                $sentence = sprintf(
+                    _("Your page was saved as 'In Progress' rather than 'Done', because you have already reached the daily page limit for %s."),
+                    $this->round->id
+                );
+                throw new DailyLimitExceededException($sentence);
+            } else {
+                $this->save_as_done();
+                if ($pre_save_dpl_count + 1 >= $this->round->daily_page_limit) {
+                    $sentence = sprintf(
+                        _("Your page has been saved as 'Done'. However, you have now reached the daily page limit for %s."),
+                        $this->round->id
+                    );
+                    throw new DailyLimitReachedException($sentence);
```

Comment:
We should probably think of something besides throwing an exception for this, because an exception is going to return a non-200 message to the client which by definition is a failure. Instead of throwing an exception this function (and endpoint) should probably return some structure that the client can parse to figure this out.

Reply:
Yes, I that is probably best.

Reply:
We could change the procedure so that the client has to `save` then `checkin`. `checkin` would return a status code and a message or throw an exception if already over the limit.

Reply:
I don't think we need to make it multi-step. We could have the API call catch these two exceptions and return something in the 200 response to tell the client.

Reply:
Or throw an exception only in the over-the-limit case. Then the client application could detect this and do a save (as in progress) which would return the normalised text as usual for such saves. The client would be behaving in a similar way to PPage.